### DISCUSSION TITLE
Create default class map and map members to a Csv Field at run-time from collection of data

### DISCUSSION THREAD

**Discussion by jlw109** [JoshClose/CsvHelper#1747] [Q&A]
Hi

I have a use-case for building ClassMap's at run-time from a collection of data that contains the property names for properties on classes used to represent data when reading or writing to csv. 

I have been building class maps at run-time for classes without nested properties. But with complex properties (i.e. properties that are other reference types I define), not sure best way to build the ClassMap at run-time from a collection of data provided at run-time. 

I put together a sample console app to demonstrate the use-case I'm trying to solve. Sample available at https://github.com/jlw109/CsvHelperDynamicClassMaps.git 

Console app main method has three methods it calls. 

            var dataFromExample1 = RunExampleWithStaticMapping();
            var dataFromExample2 = RunExampleWithStaticMappingUsingReferences();
            var dataFromExample3 = RunExampleWithDynamicMapping();

The first method called is an example with mapping the preferred way when you have nested properties and have defined a class map at compile time. 

Second method called is an example with mapping when you have nested properties and have defined a class map at compile time, but using legacy References method.

Last method called is the use-case I am trying to solve. 

See method in program.cs called BuildDynamicCsvClassMap<TModel>(). Comment explains goals of method that is incomplete. 

This relates a little to https://github.com/JoshClose/CsvHelper/discussions/1633 
### 
Update 1

I have committed changes that provides one solution to my use-case. I suspect this can be improved for reliability and performance. 

The example demonstrates how to use CsvHelper to read or write a csv file using a class map that is constructed at run time from a collection of custom objects defined by the class

    public class UserDefinedFieldForCsvMapper
    {
        public int ColumnIndex { get; set; }

        public string FieldAlias { get; set; }

        public string FieldName { get; set; }
    }

The FieldName property must match a property name on the class being used for representing the records being read or written. The property name can represent nested properties using a dot between properties (e.g. Product.Name a nested property of Transaction in the example provided).

**Comment by JoshClose** [comment]
Take a look at what `AutoMap` does. 

https://github.com/JoshClose/CsvHelper/blob/eae9c4f5a7bc9bbaa600f96626034832f662f79c/src/CsvHelper/Configuration/ClassMap.cs#L213

Depending on what you need to do, it may be useful to just call `AutoMap` and then make changes from there.

Maybe I should add a method to map by a string like `Map("A.B.C")` and let CsvHelper do the work of setting it all up. I don't know how common of a thing when doing mapping on the fly it is to have the property tree in a string like that though.
